{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "from os.path import dirname, expanduser\n",
    "sys.path.append(expanduser('~/PyRosetta4'))\n",
    "my_ip = os.environ[\"my_ip\"]\n",
    "#my_ip = '127.0.0.1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found rosetta database at: /usr/local/lib/python2.7/dist-packages/pyrosetta-2018.12+release.5ecebca-py2.7-linux-x86_64.egg/pyrosetta/database; using it....\n",
      "PyRosetta-4 2017 [Rosetta PyRosetta4.Release.python27.ubuntu 2018.12+release.5ecebca5ecebcadacdf48ede3c1981444601bf1cd47ce0d 2018-03-23T13:02:49] retrieved from: http://www.pyrosetta.org\n",
      "(C) Copyright Rosetta Commons Member Institutions.\n",
      "Created in JHU by Sergey Lyskov and PyRosetta Team.\n",
      "\n",
      "\u001b[0mcore.init: \u001b[0mChecking for fconfig files in pwd and ./rosetta/flags\n",
      "\n",
      "\u001b[0mcore.init: \u001b[0mRosetta version: PyRosetta4.Release.python27.ubuntu r174 2018.12+release.5ecebca 5ecebcadacdf48ede3c1981444601bf1cd47ce0d http://www.pyrosetta.org 2018-03-23T13:02:49\n",
      "\u001b[0mcore.init: \u001b[0mcommand: PyRosetta -ex1 -ex2aro -constant_seed -database /usr/local/lib/python2.7/dist-packages/pyrosetta-2018.12+release.5ecebca-py2.7-linux-x86_64.egg/pyrosetta/database\n",
      "\u001b[0mcore.init: \u001b[0m\u001b[31m\u001b[4mConstant seed mode\u001b[0m, seed=1111111 seed_offset=0 real_seed=1111111\n",
      "\u001b[0mcore.init.random: \u001b[0mRandomGenerator:init: Normal mode, seed=1111111 RG_type=mt19937\n"
     ]
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "import optparse    \n",
    "#from rosetta import *\n",
    "from pyrosetta import *\n",
    "init(extra_options = \"-constant_seed\")\n",
    "import os#; os.chdir('.test.output')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A GENERAL EXPLANATION\n",
    "\n",
    "folding.py\n",
    "\n",
    "This script performs fragment insertion for an input protein sequence. The\n",
    "sequence may be explicit, in a FASTA file, or in a PDB file. Two fragment files,\n",
    "preferably one longer than the other, must also be provided. Fragment insertion\n",
    "is accompanied by a Monte Carlo assessment allowing the conformation to escape\n",
    "local minima. Output structures from this protocol are intended to proceed into\n",
    "a refinement step (such as that in refinement.py) to produce reasonable\n",
    "estimates of the protein conformation.\n",
    "\n",
    "Instructions:\n",
    "\n",
    "    1) ensure that your PDB file is in the current directory\n",
    "    2) run the script:\n",
    "        from commandline                        >python D060_Folding.py\n",
    "\n",
    "        from within python/ipython              [1]: run D060_Folding.py\n",
    "\n",
    "Author: Evan H. Baugh\n",
    "    revised and motivated by Robert Schleif\n",
    "\n",
    "Updated by Boon Uranukul, 6/9/12\n",
    "Simplified special constant seed initialization ~ Labonte\n",
    "\n",
    "References:\n",
    "    P. Bradley, K. Misura, and D. Baker, \"Toward high-resolution de novo\n",
    "        structure prediction for small proteins,\" Science 309 (5742)\n",
    "        1868-1871 (2005)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### THE BASIC PROTOCOL, sample_folding\n",
    "\n",
    "This sample script is setup for usage with\n",
    "    commandline arguments,\n",
    "    default running within a python interpreter,\n",
    "    or for import within a python interpreter,\n",
    "        (exposing the sample_folding method)\n",
    "\n",
    "The method sample_folding:\n",
    "    1.  creates a pose from the desired sequence (fullatom)\n",
    "    2.  set all of the pose's backbone torsion angles\n",
    "    3.  creates a (fullatom) reference copy of the pose\n",
    "    4.  creates Movers for switching between fullatom and centroid\n",
    "    5.  convert both poses to centroid\n",
    "    6.  creates a MoveMap with all backbone torsion angles free\n",
    "    7.  sets up ClassicFragmentMovers for inserting fragments backbone\n",
    "            structures for long and short fragments\n",
    "    8.  creates RepeatMovers for the long and short fragment Movers\n",
    "    9.  create a PyMOL_Observer for viewing intermediate output\n",
    "    10. creates low and high resolution ScoreFunctions\n",
    "    11. sets up a RepeatMover on a TrialMover of a SequenceMover\n",
    "            -setup the TrialMover\n",
    "                a.  create a SequenceMover with the:\n",
    "                        >RepeatMover on the long fragment Mover\n",
    "                        >RepeatMover on the short fragment Mover\n",
    "                b.  create a MonteCarlo object for assessing moves\n",
    "                c.  create the TrialMover (on the SequenceMover)\n",
    "            -create the RepeatMover (on the TrialMover)\n",
    "    12. creates a (Py)JobDistributor for managing multiple trajectories\n",
    "    13. stores the original score evaluation (centroid)\n",
    "    14. performs low-resolution folding:\n",
    "            a. set necessary variables\n",
    "                -reload the starting (centroid) pose\n",
    "                -change the pose's PDBInfo.name, for the PyMOLMover\n",
    "                -reset the MonteCarlo object (for the TrialMover)\n",
    "            b. perform sampling and assessment using the final RepeatMover\n",
    "            c. convert the lowest scoring decoy to fullatom using the SequenceMover\n",
    "            d. export the (lowest scoring) decoy structure\n",
    "                 -recover the lowest scoring (centroid) decoy\n",
    "                 -store the decoy score\n",
    "                 -convert the decoy to fullatom\n",
    "                 -guess potential disulfide bridges\n",
    "                 -output the decoy structure using the PyJobDistributor\n",
    "                 -export the decoy structure using the PyMOL_Observer\n",
    "    15. outputs the score evaluations\n",
    "\n",
    "The second method, guess_disulfides, is called in sample_folding to print out\n",
    "cysteine residues which are close to each other in the decoy structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Methods\n",
    "\n",
    "def sample_folding(sequence,\n",
    "        long_frag_filename, long_frag_length,\n",
    "        short_frag_filename, short_frag_length,\n",
    "        kT = 3.0, long_inserts = 1, short_inserts = 3, cycles = 40,\n",
    "        jobs = 1, job_output = 'fold_output'):\n",
    "    \"\"\"\n",
    "    Performs\n",
    "        exporting structures to a PyMOL instance\n",
    "        Output structures are named  <job_output>_(job#).pdb\n",
    "    \"\"\"\n",
    "    # 1. create a pose from the desired sequence (fullatom)\n",
    "    pose = Pose()\n",
    "    # the method make_pose_from_sequence produces a complete IDEALIZED\n",
    "    #    protein conformation of the input sequence, the ResidueTypeSet (third\n",
    "    #    argument below) may be varied, and this method supports non-proteogenic\n",
    "    #    chemistry (though it is still a Rosetta Residue). however this syntax\n",
    "    #    is more involved and not robust to user errors, and not presented here\n",
    "    # small differences in bond lengths and bond angles WILL change the results,\n",
    "    #### if you desire an alternate starting conformation, alter steps\n",
    "    ####     1. and 2. as you please\n",
    "    make_pose_from_sequence(pose, sequence, 'fa_standard')\n",
    "\n",
    "    # 2. linearize the pose by setting backbone torsions to large values\n",
    "    # the method make_pose_from_sequence does not create the new pose's\n",
    "    #    PDBInfo object, so its done here, without it an error occurs later\n",
    "    pose.pdb_info(rosetta.core.pose.PDBInfo( pose.total_residue() ))\n",
    "    for i in range(1, pose.total_residue() + 1):\n",
    "        pose.set_omega(i, 180)\n",
    "        pose.set_phi(i, -150)    # reasonably straight\n",
    "        pose.set_psi(i, 150)\n",
    "        #### if you want to see the decoy scores, the PDBInfo needs these lines\n",
    "        #pose.pdb_info().chain(i, 'A')    # necessary to color by score\n",
    "        #pose.pdb_info().number(i, i)    # for PDB numbering\n",
    "    ####\n",
    "\n",
    "    # 3. create a (fullatom) reference copy of the pose\n",
    "    test_pose = Pose()\n",
    "    test_pose.assign( pose )\n",
    "    test_pose.pdb_info().name('linearized pose')\n",
    "\n",
    "    # 4. create centroid <--> fullatom conversion Movers\n",
    "    to_centroid = SwitchResidueTypeSetMover('centroid')\n",
    "    # centroid Residue objects, of amino acids, have all their sidechain atoms\n",
    "    #    replaced by a single representative \"atom\" to speed up calculations\n",
    "    to_fullatom = SwitchResidueTypeSetMover('fa_standard')\n",
    "\n",
    "    # 5. convert the poses to centroid\n",
    "    to_centroid.apply(pose)\n",
    "    to_centroid.apply(test_pose)\n",
    "\n",
    "    # 6. create the MoveMap, all backbone torsions free\n",
    "    movemap = MoveMap()\n",
    "    movemap.set_bb(True)\n",
    "    # minimizing the centroid chi angles (the sidechain centroid atoms) is\n",
    "    #    almost always USELESS since this compression is performed for speed,\n",
    "    #    not accuracy and clashes usually occur when converting to fullatom\n",
    "\n",
    "    # 7. setup the ClassicFragmentMovers\n",
    "    # for the long fragments file\n",
    "    # this \"try--except\" is used to catch improper fragment files\n",
    "    try:\n",
    "        fragset_long = core.fragment.ConstantLengthFragSet( long_frag_length , long_frag_filename )\n",
    "        #### the ConstantLengthFragSet is overloaded, this same\n",
    "        ####    ConstantLengthFragSet can be obtained with different syntax\n",
    "        # to obtain custom fragments, see Generating Fragment Files below\n",
    "    except:\n",
    "        raise IOError('Make sure long_frag_length matches the fragments in\\n\\\n",
    "            long_frag_file and that long_frag_file is valid')\n",
    "    long_frag_mover = protocols.simple_moves.ClassicFragmentMover(fragset_long, movemap)\n",
    "    # and for the short fragments file\n",
    "    # this \"try--except\" is used to catch improper fragment files\n",
    "    try:\n",
    "        fragset_short = core.fragment.ConstantLengthFragSet(short_frag_length, short_frag_filename)\n",
    "    except:\n",
    "        raise IOError('Make sure short_frag_length matches the fragments in\\n\\\n",
    "            short_frag_file and that short_frag_file is valid')\n",
    "    short_frag_mover = protocols.simple_moves.ClassicFragmentMover(fragset_short, movemap)\n",
    "\n",
    "    # 8. setup RepeatMovers for the ClassicFragmentMovers\n",
    "    insert_long_frag = protocols.moves.RepeatMover(long_frag_mover, long_inserts)\n",
    "    insert_short_frag = protocols.moves.RepeatMover(short_frag_mover, short_inserts)\n",
    "\n",
    "    # 9. create a PyMOL_Observer for exporting structures to PyMOL (optional)\n",
    "    # the PyMOL_Observer object owns a PyMOLMover and monitors pose objects for\n",
    "    #    structural changes, when changes are detected the new structure is\n",
    "    #    sent to PyMOL\n",
    "    # fortunately, this allows investigation of full protocols since\n",
    "    #    intermediate changes are displayed, it also eliminates the need to\n",
    "    #    manually apply the PyMOLMover during a custom protocol\n",
    "    # unfortunately, this can make the output difficult to interpret (since you\n",
    "    #    aren't explicitly telling it when to export) and can significantly slow\n",
    "    #    down protocols since many structures are output (PyMOL can also slow\n",
    "    #    down if too many structures are provided and a fast machine may\n",
    "    #    generate structures too quickly for PyMOL to read, the\n",
    "    #    \"Buffer clean up\" message\n",
    "    # uncomment the line below to use PyMOL_Observer\n",
    "##    AddPyMOLObserver(test_pose, True)\n",
    "\n",
    "    # 10. create ScoreFunctions\n",
    "    # for low-resolution, centroid, poses necessary for the TrialMover's\n",
    "    #    MonteCarlo object (see below)\n",
    "    scorefxn_low = create_score_function('score3')\n",
    "    # for high-resolution, fullatom, poses necessary for scoring final output\n",
    "    #    from the PyJobDistributor (see below)\n",
    "    scorefxn_high = get_fa_scorefxn() #  create_score_function('standard', 'score12')\n",
    "\n",
    "    # 11. setup a RepeatMover on a TrialMover of a SequenceMover\n",
    "    # -setup a TrialMover\n",
    "    #    a. create a SequenceMover of the fragment insertions\n",
    "    #### add any other moves you desire\n",
    "    folding_mover = protocols.moves.SequenceMover()\n",
    "    folding_mover.add_mover(insert_long_frag)\n",
    "    folding_mover.add_mover(insert_short_frag)\n",
    "    #    b. create a MonteCarlo object to define success/failure\n",
    "    # must reset the MonteCarlo object for each trajectory!\n",
    "    mc = MonteCarlo(test_pose, scorefxn_low, kT)\n",
    "    # c. create the TrialMover\n",
    "    trial = TrialMover(folding_mover, mc)\n",
    "\n",
    "    #### for each trajectory, try cycles number of applications\n",
    "\n",
    "    # -create the RepeatMover\n",
    "    folding = protocols.moves.RepeatMover(trial, cycles)\n",
    "\n",
    "    # 12. create a (Py)JobDistributor\n",
    "    jd = PyJobDistributor(job_output, jobs, scorefxn_high)\n",
    "\n",
    "    # 13. store the score evaluations for output\n",
    "    # printing the scores as they are produced would be difficult to read,\n",
    "    #    Rosetta produces a lot of verbose output when running\n",
    "    scores = [0]*(jobs + 1)\n",
    "    scores[0] = scorefxn_low(pose)\n",
    "\n",
    "    # 14. perform folding by\n",
    "    counter = 0    # for exporting to PyMOL\n",
    "    while not jd.job_complete:\n",
    "        # a. set necessary variables for the new trajectory\n",
    "        # -reload the starting pose\n",
    "        test_pose.assign(pose)\n",
    "        # -change the pose's PDBInfo.name, for the PyMOL_Observer\n",
    "        counter += 1\n",
    "        test_pose.pdb_info().name(job_output + '_' + str(counter))\n",
    "        # -reset the MonteCarlo object (sets lowest_score to that of test_pose)\n",
    "        mc.reset(test_pose)\n",
    "\n",
    "        #### if you create a custom protocol, you may have additional\n",
    "        ####    variables to reset, such as kT\n",
    "\n",
    "        #### if you create a custom protocol, this section will most likely\n",
    "        ####    change, many protocols exist as single Movers or can be\n",
    "        ####    chained together in a sequence (see above) so you need\n",
    "        ####    only apply the final Mover\n",
    "        # b. apply the refinement protocol\n",
    "        folding.apply(test_pose)\n",
    "\n",
    "        ####\n",
    "        # c. export the lowest scoring decoy structure for this trajectory\n",
    "        # -recover the lowest scoring decoy structure\n",
    "        mc.recover_low(test_pose)\n",
    "        # -store the final score for this trajectory\n",
    "        scores[counter] = scorefxn_low(test_pose)\n",
    "        # -convert the decoy to fullatom\n",
    "        # the sidechain conformations will all be default,\n",
    "        #    normally, the decoys would NOT be converted to fullatom before\n",
    "        #    writing them to PDB (since a large number of trajectories would\n",
    "        #    be considered and their fullatom score are unnecessary)\n",
    "        # here the fullatom mode is reproduced to make the output easier to\n",
    "        #    understand and manipulate, PyRosetta can load in PDB files of\n",
    "        #    centroid structures, however you must convert to fullatom for\n",
    "        #    nearly any other application\n",
    "        to_fullatom.apply(test_pose)\n",
    "        # -guess what cysteines are involved in disulfide bridges\n",
    "        guess_disulfides(test_pose)\n",
    "        # -output the fullatom decoy structure into a PDB file\n",
    "        jd.output_decoy(test_pose)\n",
    "        # -export the final structure to PyMOL\n",
    "        test_pose.pdb_info().name(job_output + '_' + str(counter) + '_fa')\n",
    "\n",
    "        #### if you want to see the decoy scores, uncomment the line below\n",
    "        #scorefxn_high( test_pose )\n",
    "\n",
    "    # 15. output the score evaluations\n",
    "    print( '===== Centroid Scores =====' )\n",
    "    print( 'Original Score\\t:\\t', scores[0] )\n",
    "    for i in range(1, len(scores)):    # print out the job scores\n",
    "        # the \"[:14].ljust(14)\" is to force the text alignment\n",
    "        print( (job_output + '_' + str( i ))[:14].ljust(14) +\\\n",
    "            '\\t:\\t', scores[i] )\n",
    "\n",
    "    return scores    # for other protocols\n",
    "\n",
    "# locates all cysteines, make bonds between them, output the bonds that lower\n",
    "#    the score\n",
    "def guess_disulfides(pose, cutoff = 6.0):\n",
    "    \"\"\"\n",
    "    A quick method for probing a protein for cysteine residues close to each\n",
    "        other (within  <cutoff>  )\n",
    "    \"\"\"\n",
    "    # find all cysteine residues and consider possible disulfides\n",
    "    cys = [i for i in range(1, pose.total_residue() + 1)\n",
    "        if pose.residue(i).name1() == 'C']\n",
    "    partners = [0]*sum( range(len(cys)) )    # all disulfides possible\n",
    "    i = 0\n",
    "    # create all combinations\n",
    "    for first in range(len(cys[:-1])):\n",
    "        for second in cys[first + 1:]:\n",
    "            partners[i] = (cys[first], second)\n",
    "            i += 1\n",
    "    # try each disulfide, if it lowers the score, print it to screen\n",
    "    print( '='*80 )\n",
    "    print( 'Potential Disulfides:' )\n",
    "    for pair in partners:\n",
    "        # for a fullatom cysteine in PyRosetta, the 6th atom is sulfur\n",
    "        separation = (pose.residue(pair[0]).xyz(6) -\n",
    "            pose.residue(pair[1]).xyz(6)).norm\n",
    "        if separation < cutoff:\n",
    "            print( 'between (pose numbered) residue' , pair[0] , 'and' ,\\\n",
    "                pair[1] , '|' , separation , 'Angstrom separation' )\n",
    "    print( '='*80 )\n",
    "    # to manipulate disulfide bonds, use:\n",
    "    # formation:    form_disulfide(pose.conformation(), 6, 16)\n",
    "    # cleavage:     change_cys_state(6, 'CYS', pose.conformation() )\n",
    "    #               change_cys_state(16, 'CYS', pose.conformation() )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### INTERPRETING RESULTS\n",
    "\n",
    "The (Py)JobDistributor will output the lowest scoring pose for each trajectory\n",
    "(as a PDB file), recording the score in <job_output>.fasc. Generally,\n",
    "the decoy generated with the lowest score contains the best prediction\n",
    "for the protein conformation. In this case the scores may be misleading since\n",
    "the output PDB files were converted into fullatom mode. Centroid scores are\n",
    "written to the standard output (the screen) but not to file. If you need these\n",
    "scores, you must reload the poses, convert them to centroid, and rescore them.\n",
    "Using low sampling with an average acceptance (kT), many of the structures will\n",
    "appear incomplete (not fully folded) and unrealistic (clashes). Fragment\n",
    "insertion efficiently reproduces secondary structural elements in the proposed\n",
    "fold. If this method is followed by refinement, a higher kT may be useful since\n",
    "this will produce compact centroid folds that have clashes easily removed during\n",
    "refinement. Minimization steps can be confounded by clashes but this approach\n",
    "yields better fold predictions.\n",
    "\n",
    "For each trajectory, the PyMOLMover will export numerous intermediate centroid\n",
    "conformations and the final fullatom fold. Depending on the machine used,\n",
    "the structures may be produced too quickly for PyMOL to display properly.\n",
    "If this occurs, some structures will NOT load into PyMOL and a message will\n",
    "print in the PyMOL upper window indicating \"Buffer clean up\". Please change\n",
    "where the PyMOLMover is applied to view different output or change the method\n",
    "to include pauses (try the Python time module, specifically time.sleep). These\n",
    "alterations are tedious but provide an effective means for tuning the protocol\n",
    "parameters while maintaining biochemical feasibility (i.e. you can know if the\n",
    "fold is wrong).\n",
    "\n",
    "The method guess_disulfides prints out pairs of cysteine residues in the input\n",
    "pose which are within a certain distance (default 6 Angstroms). The method\n",
    "sample_folding calls guess_disulfides for the final decoy of each trajectory.\n",
    "To manipulate (fullatom) disulfide bonds in PyRosetta, use the commands:\n",
    "form_disulfide and change_cys_state (see the comments at the end of\n",
    "guess_disulfides)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-6-9e52ef75b172>, line 4)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-6-9e52ef75b172>\"\u001b[0;36m, line \u001b[0;32m4\u001b[0m\n\u001b[0;31m    short_frag_filename =\u001b[0m\n\u001b[0m                          ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "sequence = \"DAITIHSILDWIEDNLESPLSLEKVSERSGYSKWHLQRMFKKETGHSLGQYIRSRKMTEIAQKLKESNEPILYLAERYGFESQQTLTRTFKNYFDVPPHKYRMTNMQGESRFLHPL\"\n",
    "long_frag_filename = \"test_in.pdb\"\n",
    "long_frag_length = 116\n",
    "short_frag_filename = \n",
    "short_frag_length = \n",
    "sample_folding(sequence,\n",
    "        long_frag_filename, long_frag_length,\n",
    "        short_frag_filename, short_frag_length,\n",
    "        kT = 3.0, long_inserts = 1, short_inserts = 3, cycles = 40,\n",
    "        jobs = 1, job_output = 'fold_output')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# COMMANDLINE COMPATIBILITY\n",
    "\n",
    "# everything below is added to provide commandline usage,\n",
    "#   the available options are specified below\n",
    "# this method:\n",
    "#    1. defines the available options\n",
    "#    2. loads in the commandline or default values\n",
    "#    3. calls sample_folding with these values\n",
    "\n",
    "# parser object for managing input options\n",
    "# all defaults are for the example using \"test_in.pdb\" with reduced\n",
    "#    cycles/jobs to provide results quickly\n",
    "\n",
    "\n",
    "parser = optparse.OptionParser()\n",
    "\n",
    "parser.add_option( '--pdb_filename', dest = 'pdb_filename',\n",
    "    default = '../test/data/test_in.pdb',    # default example PDB\n",
    "    help = 'the PDB file containing the protein to fold')\n",
    "parser.add_option('--fasta_filename', dest = 'fasta_filename',\n",
    "    default = '',    # default empty!\n",
    "    help = 'the FASTA file containing the protein to sequence to fold')\n",
    "parser.add_option( '--sequence', dest = 'sequence',\n",
    "    default = '',    # default empty!\n",
    "    help = 'the protein sequence to fold')\n",
    "# the fragment files options\n",
    "parser.add_option('--long_frag_filename', dest = 'long_frag_filename',\n",
    "    default = '../test/data/test9_fragments',    # specific to each PDB (test_in.pdb here)\n",
    "    help = 'the file of long fragments corresponding to the sequence')\n",
    "parser.add_option('--long_frag_length', dest = 'long_frag_length',\n",
    "    default = '9',    # must match the long_frag_filename\n",
    "    help = 'the length of fragments contained in the long_frag_file')\n",
    "parser.add_option('--short_frag_filename', dest = 'short_frag_filename',\n",
    "    default = '../test/data/test3_fragments',    # specific to each PDB (test_in.pdb here)\n",
    "    help = 'the file of short fragments corresponding to the sequence')\n",
    "parser.add_option( '--short_frag_length', dest = 'short_frag_length',\n",
    "    default = '3',    # must match the short_frag_filename\n",
    "    help = 'the length of fragments contained in the short_frag_file')\n",
    "# folding protocol options\n",
    "parser.add_option('--kT', dest='kT',\n",
    "    default = '3.0',    # default higher for easy demonstration\n",
    "    help = 'the \\\"temperature\\\" of the sample folding protocol')\n",
    "parser.add_option('--long_inserts', dest='long_inserts',\n",
    "    default = '1',\n",
    "    help = 'the number of times a long fragment insertion is applied per\\\n",
    "        cycle in the sample folding protocol')\n",
    "parser.add_option('--short_inserts', dest='short_inserts',\n",
    "    default = '3',\n",
    "    help = 'the number of times a short fragment insertion is applied per\\\n",
    "        cycle in the sample folding protocol')\n",
    "parser.add_option('--cycles', dest='cycles',\n",
    "    default = '40' ,\n",
    "    help = 'the number of folding rounds (long and short fragment\\\n",
    "        insertions) in the sample folding protocol')\n",
    "# PyJobDistributor options\n",
    "parser.add_option('--jobs', dest='jobs',\n",
    "    default = '1',    # default to single trajectory for speed\n",
    "    help = 'the number of jobs (trajectories) to perform' )\n",
    "parser.add_option('--job_output', dest = 'job_output',\n",
    "    default = 'fold_output',    # if a specific output name is desired\n",
    "    help = 'the name preceding all output, output PDB files and .fasc')\n",
    "(options,args) = parser.parse_args()\n",
    "\n",
    "# the user may input a PDB file, fasta file, or sequence directly\n",
    "# PDB file option\n",
    "pose = Pose()\n",
    "# Fasta file option\n",
    "fasta_filename = options.fasta_filename\n",
    "if fasta_filename:    # defaults to off, empty string\n",
    "    f = open(fasta_filename, 'r')    # open the file\n",
    "    sequence = f.readlines()    # read the text\n",
    "    f.close()    # close it\n",
    "    # removing the trailing \"\\n\" and any header lines\n",
    "    sequence = [line.strip() for line in sequence if not '>' in line]\n",
    "    sequence = ''.join( sequence )    # combine into a single sequence\n",
    "elif options.sequence:\n",
    "    sequence = options.sequence\n",
    "else:\n",
    "    pdb_filename = options.pdb_filename; #Default is the test PDB, not an empty string.\n",
    "    pose_from_file(pose, pdb_filename)\n",
    "    sequence = pose.sequence()\n",
    "#Checks for the sequence in a fasta, then direct, and finally from a PDB file.  If no PDB file is given, it will load the default.\n",
    "\n",
    "# fragment files options\n",
    "long_frag_filename = options.long_frag_filename\n",
    "long_frag_length = int(options.long_frag_length)\n",
    "short_frag_filename = options.short_frag_filename\n",
    "short_frag_length = int(options.short_frag_length)\n",
    "# folding protocol options\n",
    "kT = float(options.kT)\n",
    "long_inserts = int(options.long_inserts)\n",
    "short_inserts = int(options.short_inserts)\n",
    "cycles = int(options.cycles)\n",
    "# PyJobDistributor options\n",
    "jobs = int(options.jobs)\n",
    "job_output = options.job_output\n",
    "\n",
    "# ''' Disabled until Abinitio error is fixed\n",
    "sample_folding(sequence,\n",
    "    long_frag_filename, long_frag_length,\n",
    "    short_frag_filename, short_frag_length,\n",
    "    kT, long_inserts, short_inserts, cycles,\n",
    "    jobs, job_output)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ALTERNATE SCENARIOS\n",
    "\n",
    "### Obtaining Sequences\n",
    "\n",
    "PDB files are essential to all Rosetta applications. Occasionally protein\n",
    "sequences are required instead. In many cases, only proteins with solved\n",
    "crystal structures are used and as such, sequences, as with PDB files, are\n",
    "obtained from the RCSB website. To obtain the protein sequence from a PDB file:\n",
    "\n",
    "    1) locate your protein of interest at http://www.pdb.prg/\n",
    "    2) download the PDB file, using a browser this includes:\n",
    "        a. clicking \"Download Files\" on the upper right\n",
    "        b. clicking \"FASTA Sequence\", the second option\n",
    "\n",
    "### FASTA Files, PDB Files, and Biopython\n",
    "\n",
    "The FASTA file can be input to this method using the --fasta_filename option.\n",
    "Most FASTA files, including those obtained from RCSB, will start with a header\n",
    "line staring with the \">\" character. FASTA files from RCSB may contain multiplt\n",
    "chains. Rosetta is designed to predict small globular domain folds and will\n",
    "not work effectively with multiple chains.\n",
    "Using PyRosetta, a PDB file protein sequence can be extracted from a Pose object\n",
    "using the method Pose.sequence.\n",
    "Biopython is an effective tool for extracting information from biological data\n",
    "files, including FASTA files and PDB files. Since it is in Python, it is a very\n",
    "useful tool for combining PyRosetta structural analysis and protocols with\n",
    "sequence level analysis or quick calculations.\n",
    "\n",
    "\n",
    "\n",
    "### A Real Example\n",
    "\"\"\"\n",
    "All of the default variables and parameters used above are specific to\n",
    "the example with \"test_in.pdb\", which is supposed to be simple,\n",
    "straightforward, and speedy. Here is a more practical example:\n",
    "\n",
    "Pheromone ER-23 is a small protein (51 amino acids) with 5 disulfide bridges.\n",
    "The formation of these bonds may significantly effect folding. Suppose you\n",
    "are interested in the significance of these disulfide bonds as folding\n",
    "constraints and want to predict the protein without them using PyRosetta.\n",
    "\n",
    "    1. Obtain the protein sequence of RCSB PDB 1HA8 (instructions above)\n",
    "            (for the example here, download the PDB file and ensure it is clean)\n",
    "    2. Make two fragment files using the \"Generate Fragment Files\"\n",
    "            (instructions above)\n",
    "            -of 9-mers (the long fragments)\n",
    "            -of 3-mers (the short fragments)\n",
    "    3. Make a directory containing:\n",
    "            -the PDB file for 1HA8 (cleaned of HETATMs and waters)\n",
    "                lets name it \"1HA8.clean.pdb\" here\n",
    "            -the 9-mer fragment file for 1HA8\n",
    "                lets name it \"1HA8.frag9\" here\n",
    "            -the 3-mer fragment file for 1HA8\n",
    "                lets name it \"1HA8.frag3\" here\n",
    "            -this sample script (technically not required, but otherwise the\n",
    "                commands in 4. would change since folding.py would not be here)\n",
    "    4. Run the script from the commandline with appropriate arguments:\n",
    "\n",
    ">python folding.py --pdb_filename 1HA8.clean.pdb --long_frag_filename 1HA8.frag9 --long_frag_length 9 --short_frag_filename 1HA8.frag3 --short_frag_length 3 --jobs 100 --job_output 1HA8_folding_output --kT 1.0 --long_inserts 1 --short_inserts 3 --cycles 200 --disulfides 1\n",
    "\n",
    "        -The long_frag_length MUST match long_frag_filename (9 in this example)\n",
    "        -The short_frag_length MUST match short_frag_filename (3 here)\n",
    "        -100 trajectories is low, sampling protein conformations requires many\n",
    "            trials, typically hundreds (800-1000) trajectories are attempted\n",
    "        -The \"temperature\" parameter, kT, is set to 1.0, a neutral value,\n",
    "            larger kT increase the diversity of sampling (easily escape local\n",
    "            minima) but requires compensation with more trajectories\n",
    "            smaller kT decreases the chance of sampling \"useless\" space but\n",
    "            cannot easily escape local minima (the default kT = 3.0\n",
    "        -There are no common values for long_inserts, short_inserts, or cycles\n",
    "            more inserts indicates greater sampling with fragments while\n",
    "            greater cycles produces more sampling in general (though in this\n",
    "            case only fragment insertion is performed)\n",
    "            however, only one selection (Metropolis Criteria) is applied per\n",
    "            cycle, thus a proposed structure with too many fragments inserted\n",
    "            is likely to be rejected but additional trials increases the\n",
    "            computation time noticeably\n",
    "            here, as is typical is Rosetta, a fairly conservative change is\n",
    "            proposed (1 9-mer and 3 3-mers) per cycle with many cycles\n",
    "\n",
    "    5. Wait for output, this will take a while (performing 100 trajectories\n",
    "            of loop modeling involving 1*3*200 total moves per trajectory)\n",
    "    6. Analyze the results (see INTERPRETING RESULTS above)\n",
    "\n",
    "Note: this is NOT intended to be used for realistic centroid folding,\n",
    "it merely provides a \"skeleton\" for the code in PyRosetta. It may be useful\n",
    "for preliminary investigation but the best protocols are somewhat\n",
    "protein-specific, there is no current universal folding method. As mentioned\n",
    "above, this script only touches the low-resolution aspect of Rosetta folding\n",
    "and must be accompanied by a high-resolution step to yield complete results.\n",
    "Generally, a protocol similar to the one presented here with more drastic\n",
    "sampling and a larger number of trials should be sufficient to find realistic\n",
    "folds (after refinement).\n",
    "\n",
    "\n",
    "### Changing Folding Sampling\n",
    "\n",
    "Rosetta is a structure prediction tool and all predictions rely on the\n",
    "techniques (moves) used to generate proposed structures. There is NO GENERAL\n",
    "SOLUTION for predicting a protein fold from scratch and fragment insertion\n",
    "is an effective way of sampling the relevant space. Since fragments are\n",
    "sequence and secondary structure (prediction) dependent, the resultant folds\n",
    "are biochemically feasible and in accordance with our prior observations about\n",
    "protein stucture (the PDB).\n",
    "\n",
    "For speed, Rosetta folding (and the folding algorithm presented here) is\n",
    "performed in low-resolution (centroid) mode. This method of folding is\n",
    "abstractly looking for structures which obey certain size constraints (i.e. the\n",
    "rg score term) but do not unrealistically clash (i.e. the vdw score term) that\n",
    "are achieved easily (i.e. fragment insertion, a fast algorithm). Sufficient\n",
    "sampling and downstream refinement will convert these general folds (the output\n",
    "of this script) into reasonable estimates of the native protein structure.\n",
    "\n",
    "Many other sampling techniques and methods are available. PyRosetta was designed\n",
    "to make novel investigation of these ideas easy. Recommend literature include\n",
    "topics in protein structure prediction, Monte Carlo Markov Chain processes,\n",
    "Rosetta protocols, and simulated annealing algorithms. When developing custom\n",
    "protocols, you MUST understand the Movers available in PyRosetta (see the other\n",
    "sample scripts, specifically refinement.py). Unfortunately, you cannot build\n",
    "novel Mover classes in PyRosetta. Fortunately, it is easy to implement similar\\\n",
    "code on your own once you understand how moves and selection are performed.\n",
    "\n",
    "Please try alternate sampling methods to better understand how these\n",
    "algorithms perform and to find what moves best suite your problem.\n",
    "\n",
    "\n",
    "\n",
    "### Changing Folding Scoring\n",
    "\n",
    "The Rosetta \"force-field\" is a biasing function (see MCMC algorithms) used to\n",
    "predict protein structures based on what is observed about other protein\n",
    "structures. These scoring methods supply the constraints necessary to bias\n",
    "structure prediction but incorporate a LARGE number of fine tuned parameters.\n",
    "There is no easy way to optimize all relevant constants and performance has been\n",
    "increased by balancing speed and accuracy.\n",
    "\n",
    "Many other relevant metrics exist for protein structures and discriminating\n",
    "decoys (incorrect predictions) from actual proteins is an ongoing task. Please\n",
    "experiment with different ScoreTypes in PyRosetta, their weights, and the\n",
    "combination of terms. Unfortunately, you cannot build novel ScoreTypes (scoring\n",
    "methods) in PyRosetta. Fortunately, it is easy to implement similar code on your\n",
    "own once you understand how scoring biases MCMC simulations.\n",
    "\n",
    "Please try alternate scoring functions or unique selection methods to better\n",
    "understand which scoring terms contribute to performance and to find what\n",
    "scoring best suites your problem.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
